discovery/swarmkit/swarmkit.go

```diff
@@ -0,0 +1,162 @@
+package swarmkit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/discovery"
+	"github.com/docker/swarm/cluster"
+)
+
+const defaultAgentPort = "2375"
+
+// Discovery is exported.
+type Discovery struct {
+	heartbeat  time.Duration
+	ttl        time.Duration
+	httpClient *http.Client
+	apiClient  *client.Client
+	uri        string
+	token      string
+	agentPort  int
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("swarmkit", &Discovery{})
+}
+
+func (s *Discovery) SetAgentPort(discoveryEnginePort int) {
+	s.agentPort = discoveryEnginePort
+}
+
+// Initialize initializes the discovery with the SwarmKit manager's address
+func (s *Discovery) Initialize(uri string, heartbeat time.Duration, ttl time.Duration, _ map[string]string) error {
+	// the uri may be a manager address or a docker socket
+	if uri == "" {
+		return errors.New("Swarm Mode manager address not provided")
+	}
+	s.uri = uri
+	s.heartbeat = heartbeat
+	s.ttl = ttl
+
+	// create a client to connect with the Swarm Mode API
+	httpClient, _, err := cluster.NewHTTPClientTimeout("tcp://"+uri, nil, time.Duration(30*time.Second), nil)
+	if err != nil {
+		return err
+	}
+	s.httpClient = httpClient
+
+	apiClient, err := client.NewClient("tcp://"+uri, "", s.httpClient, nil)
+	if err != nil {
+		return err
+	}
+	s.apiClient = apiClient
+
+	return nil
+}
+
+// fetch returns the list of entries for the discovery service at the specified endpoint.
+func (s *Discovery) fetch() (discovery.Entries, error) {
+	ctx := context.Background()
+	nodeList, err := s.apiClient.NodeList(ctx, types.NodeListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to fetch entries, Discovery service returned %s", err.Error())
+	}
+
+	var addrs []string
+
+	for _, n := range nodeList {
+		// report all nodes that are not in drain mode
+		if n.Spec.Availability != swarm.NodeAvailability("drain") {
```

Comment:
I'm not sure whether we should exclude nodes that are on Drain from the node inventory entirely. This could lead to mis-reporting of the inventory and operational confusion. If we want to exclude "drain" nodes from scheduling, we should consider adding a separate mechanism that abstracts this behavior out. For example, a node could be marked as "excluded" in the memory of the swarm manager, and a scheduling filter can prevent scheduling on excluded nodes. This will still allow us to view resources present on that node and consider it part of the inventory. The `Watch` method below could make sure to mark/unmark excluded nodes as that information becomes available through the events stream.

Reply:
I think it really depends on what kind of usage we wish to expose with the inventory. I see your point that it's worthwhile giving users full access to prevent confusion. Do you also think it's not the job of Swarm classic to be respecting drain constraints, since this is not a Swarm classic abstraction per se?

Reply:
I hope that is okay, that I add my opinion here. I am really waiting for this feature. Our use case is that we run a custom healtchecker periodicaly over all running containers to see if they are still up to reschedule them if a node or a process fails. But for rolling node updates it is really cool to mark a node as draining. If the node drops immediately when it is marked as draining our healtchecker will reschedule all containers immediately. That's really bad for us since we do not have a phase where we actually can drain the node (rescheduling all running containers).

Reply:
@fbuecklers thanks for sharing your use case. Sounds like you would also benefit if we don't make assumptions around drain nodes. I've removed this check from this PR.